### ISSUE TITLE
Rust HRTB breaks smart indent

### ISSUE THREAD

**Issue by kawadakk** [helix-editor/helix#7695]
### Summary

Smart indent misbehaves when editing a Rust source file containing [RPIT](https://doc.rust-lang.org/1.71.0/reference/types/impl-trait.html#abstract-return-types) with [HRTB](https://doc.rust-lang.org/1.71.0/reference/trait-bounds.html#higher-ranked-trait-bounds).

### Reproduction Steps

 

Execute <kbd>o</kbd> (open new line below selection) with a single cursor located anywhere inside line 1:

```rust
fn foo() -> impl for<'a> PartialEq<&'a str> {
    std::process::abort() as &str
}
```

I expected this to happen: The new line is pre-indented to match the indentation of the existing statement.

Instead, this happened: The new line is not pre-indented, and the cursor moves to the leftmost column.

### Helix log

~/.cache/helix/helix.log: (no relevant outputs found)

### Platform

Linux

### Terminal Emulator

tmux / Wez's Terminal

### Helix Version

helix 23.05

**Comment by pascalkuthe** [comment]
this is an upstream issue it seems the rust grammar can't parse this code correctly and emits error nodes (it doern't even recognize this as a function). Sadly TS parsers are much less error resistant than ideal (RA approach to paring is much more robust but oh well). This will break all TS functionality as a result

**Comment by CalebLarsen** [comment]
@kawadakk can you confirm if this issue is still happening? My local attempts are showing that the problem has been resolved at some point?

**Comment by the-mikedavis** [comment]
Yeah an update of tree-sitter-rust must've fixed this. I see the example code parsing correctly now, no ERROR nodes
